File: soap/ruby/st-client.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*- vim:fileencoding=utf-8:

require 'socialtext/soap'
require 'getoptlong'

DefaultOpts = {
    :wsdl => 'https://www.socialtext.net/static/wsdl/0.9.wsdl',
    :username => 'soap.test@socialtext.com',
    :password => 'bubbles',
    :workspace => 'st-soap-test',
    :page_name => 'Soap Test',
    :other_user => 'devnull11@socialtext.com'
}

getopt = GetoptLong.new(
    [ '--workspace',  '-W', GetoptLong::OPTIONAL_ARGUMENT ],
    [ '--wsdl',       '-w', GetoptLong::OPTIONAL_ARGUMENT ],
    [ '--username',   '-u', GetoptLong::OPTIONAL_ARGUMENT ],
    [ '--password',   '-p', GetoptLong::OPTIONAL_ARGUMENT ],
    [ '--page-name',  '-P', GetoptLong::OPTIONAL_ARGUMENT ],
    [ '--other-user', '-o', GetoptLong::OPTIONAL_ARGUMENT ],
    [ '--help',       '-h', GetoptLong::NO_ARGUMENT ]
)

opts = DefaultOpts
getopt.each do |key,value|
    key.gsub!("--", "");
    key.gsub!("-", "_");
    opts[key.to_sym] = value
end

if opts[:help]
    puts <<USAGE 
USAGE ./st-client.rb [OPTIONS]
OPTIONS:
  --workspace,-W  Set workspace (default '#{DefaultOpts[:workspace]}')
  --wsdl,-w       Set URL to wsdl (default '#{DefaultOpts[:wsdl]}')
  --username,-u   Set username (default '#{DefaultOpts[:username]}')
  --password,-p   Set password (default '#{DefaultOpts[:password]}')
  --page-name,-P  Set page to access (default '#{DefaultOpts[:page_name]}')
  --other-user,-u Set user to act_as (default '#{DefaultOpts[:other_user]}')
  --help,-h       This text.
USAGE
    exit 1
end

class MyClient
    include Socialtext::Soap::Client
    def initialize(*opts)
        self.st_init(*opts)
    end
end

st = MyClient.new(opts)
page_name = opts[:page_name]
other_user = opts[:other_user]

# Test the Heartbeat
puts "=== HEARTBEAT ==="
puts st.heartbeat()

# Get page
puts
puts "=== GET PAGE " + page_name + " ==="
puts st.get_page(:page => page_name, :format => 'wikitext').pageContent

# Set page
puts
puts "=== SET PAGE " + page_name + " ==="
page = st.set_page(:page => page_name, :content => 'this is tensegrity')
puts page.pageContent

# Make a new page
# Note that ruby pays no attention to the encoding set in your locale, and
# neither do we.  We assume here that your terminal expects UTF-8.
NEW_PAGE_NAME = '99¢ Veggieburger'
NEW_PAGE_BODY = <<'EOF'
At McSOAPie's™, you can get a Veggieburger™ for only 99¢.
EOF
puts
puts "=== MAKE #{NEW_PAGE_NAME} ===";
puts st.set_page(:page => NEW_PAGE_NAME, :content => NEW_PAGE_BODY).pageContent

# sleep a while to let indexing complete
sleep 10 

# Search for the changes
puts
puts "=== SEARCH tensegrity ==="
st.get_search(:query => 'tensegrity').each { |item|
    puts "#{item.subject} #{item.author} #{item.date}"
}

# Recent Changes
puts
puts "=== RECENT CHANGES ==="
st.get_changes(:count => 2).each { |item|
    puts "#{item.subject} #{item.author} #{item.date}"
}

# Set page as someone else
puts
puts "=== SET PAGE " + page_name + " as " + other_user + " ==="
st.st_act_as = other_user
puts st.set_page(:page => page_name, :content => 'I like cows').pageContent

File: soap/python/st-client.py
#!/usr/bin/env python
import sys
import SOAPpy
import time

from optparse import OptionParser

class SocialtextClient:

    def __init__(self, wsdl, *args):
        self.tokens = {}
        self.act_as = None
        self.connect_to_server(wsdl)
        if len(args) > 0:
            self.set_default_auth(*args)

    def connect_to_server(self, wsdl):
        self.wsdl = wsdl
        self.soap = SOAPpy.WSDL.Proxy(wsdl or self.wsdl)

    def set_default_auth(self, user, password, ws, act_as=None):
        self.username = user
        self.password = password
        self.workspace = ws
        self.default_token =  self.get_auth(user, password, ws, act_as)

    def act_as_default(self):
        self.act_as = None

    def act_as_user(self, act_as):
        self.act_as = act_as
        self.tokens[act_as] = self.get_auth(self.username,
                                            self.password,
                                            self.workspace,
                                            act_as)

    def current_token(self):
        if self.act_as:
            return self.tokens[self.act_as]
        else:
            return self.default_token

    def get_auth(self, *args):
        return self.soap.getAuth(*args)

    def heartbeat(self):
        return self.soap.heartBeat()

    def get_page(self, page, format='wikitext'):
        return self.soap.getPage(self.current_token(), page)

    def get_changes(self, num=4, name='recent changes'):
        changes = self.soap.getChanges(self.current_token(), name, num)
        return changes['pageMetadata']

    def get_search(self, query):
        result = self.soap.getSearch(self.current_token(), query)
        return result['pageMetadata']

    def set_page(self, page, content):
        result = self.soap.setPage(self.current_token(), page, content)
        return result

if __name__ == '__main__':
    NEW_PAGE_NAME = u'99\N{CENT SIGN} Snakeburger';
    NEW_PAGE_BODY = (u"At McSOAPie's\N{TRADE MARK SIGN}, "
                     u"you can get a Snakeburger\N{TRADE MARK SIGN} "
                     u"for only 99\N{CENT SIGN}.")

    parser = OptionParser()
    parser.add_option("-W", "--workspace",
            help="The Workspace to address")
    parser.add_option("-w", "--wsdl",
            default="https://www.socialtext.net/static/wsdl/0.9.wsdl",
            help="The URL of the WSDL")
    parser.add_option("-u", "--username",
            help="Username to use the service")
    parser.add_option("-p", "--password",
            help="The password of the username")
    parser.add_option("-o", "--other-user", dest="act_as",
            help="The name of the user to impersonate if the service")
    (options, args) = parser.parse_args()

    # Get command line arguments
    wsdl = options.wsdl
    auth = (options.username, options.password, options.workspace,
            options.act_as)

    # Create a Socialtext Client and do heartbeat
    st = SocialtextClient(wsdl, *auth)

    # Heartbeat
    print "=== HEARTBEAT ==="
    print st.heartbeat()

    print "\n=== MAKE PAGE %s ===" % NEW_PAGE_NAME
    print st.set_page(NEW_PAGE_NAME, NEW_PAGE_BODY)['pageContent']

    # Get page
    print "\n=== GET PAGE admin_wiki ==="
    print st.get_page('admin_wiki')['pageContent']

    # Set page
    print "\n=== SET PAGE admin_wiki ==="
    print st.set_page('admin_wiki', 'this is tensegrity')['pageContent']
    time.sleep(10)  # Give time for page to get indexed

    # Search for the changes we just added
    print "\n=== SEARCH tensegrity ==="
    for result in st.get_search("tensegrity"):
        print "%s - %s - %s" % (result['subject'], result['author'], 
                                result['date'])

    # Recent changes
    print "\n=== RECENT CHANGES ==="
    for change in st.get_changes():
        print "%s - %s - %s" % (change['subject'], change['author'], 
                                change['date'])

    # Set page as someone else
    print "\n=== SET PAGE cows_are_good AS devnull2@socialtext.com ==="
    st.act_as_user('devnull2@socialtext.com')
    print st.set_page('cows_are_good', 'I like cows')['pageContent']

    # Recent changes
    print "\n=== RECENT CHANGES  ==="
    for change in st.get_changes():
        print "%s - %s - %s" % (change['subject'], change['author'], 
                                change['date'])
